(android) Add unit tests for run and retryPromise

Menardi opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected
Android

### What does this PR do?
I noticed the test coverage for `cordova-android` was rather low, so I decided to start writing some tests. To figure out the test system, I first covered `retry.js`, which is a simple module that retries a promise-based function a given number of times. I also fixed a typo in this file which was mildly annoying.

The main tests I have written are in `run.js`, which now has 100% test coverage (up from 27%). This module handles the selection of the device or emulator to run on.

I would like to continue to add unit tests to help improve the code coverage, but am submitting this PR first to check that I am on the right track. I have used ES2015 notation for the tests to keep it more readable - I don't anticipate this to be a problem since it is supported in all recent node versions.  

### What testing has been done on this change?
I have run the tests using `npm run unit-tests`, and checked the coverage with `npm run cover`.

### Checklist
- [ ] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [ ] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [x] Added automated test coverage as appropriate for this change.

Menardi commented:
It looks like the tests fail on Appveyor because I am using ES2015 features and that the Appveyor set up includes Node 4 (as well as 6 and 8). I suggest removing Node 4 from there and adding 10. Would that be OK?

dpogue commented:
Yes, node 4 is no longer supported so we're slowly going through all the repos and dropping it from CI while adding node 10.

I think that's what we should do here, but please do it as its own pull request to avoid this one getting cluttered with unrelated changes :)

Menardi commented:
@dpogue Since #442 was merged (which removed support for Node 4), these tests now pass. I think this is now ready to merge :+1:

raphinesse commented:
Great PR :+1: 

There is one issue though: the handling of promises in tests. All async tests should return a promise instead of using the done callback. That way, Jasmine automatically handles resolution and rejection of these promises. Right now, failing tests would time out and you wouldn't get to see the error that caused the rejection.

Thanks a lot for the contribution!

Menardi commented:
Thanks for the suggestion, I didn't realise that was possible. According to the [Jasmine async docs](https://jasmine.github.io/tutorials/async), support for returning promises was added in 2.7, but cordova-android uses 2.6. I'll look into upgrading and see if it causes any issues.

Menardi commented:
@raphinesse I updated Jasmine to the latest version (3.1.0) without any issues. I know it is suggested to squash commits in this repo, but I put the updating of this version in a second commit as I think it makes sense to keep it separate.

As for the tests themselves, I have rewritten them to return the promises instead of calling done. There are some asynchronous tests where calling done is needed (such as testing that the promise is rejected). For those tests, I have added handling for the failure cases to ensure that all the tests fail in a meaningful way (that is, avoiding timeouts where possible).

raphinesse commented on spec/unit/run.spec.js:
```diff
@@ -17,22 +17,217 @@
+        it('should print out usage and help', () => {
+            let message = '';
+            spyOn(console, 'log').and.callFake(msg => { message += msg; });
+            spyOn(process, 'exit');
```
This clobbers the global `console.log` and `process.exit` for the whole process. I'm not too comfortable with that. Could cause very obscure errors for other tests.

Menardi replied:
I was worried about that when writing this, and looked into how to unspy a function in Jasmine, but it seems that Jasmine automatically unspies after a test is run. So, `console.log` and `process.exit` would not be spied in other functions. Personally I would prefer to explicitly unspy to avoid this confusion, but from what I can find, there is no way to do that.

raphinesse replied:
Interesting, I did not know that spies get unspied. OTOH it's a good thing, but OTOH I agree that explicit operations are preferable at times. Conveys intent and informs unsuspecting users like me :sweat_smile:

raphinesse commented:
Another note: when we finally have unbundled dependencies, we probably should replace `retryPromise` with [p-retry](https://github.com/sindresorhus/p-retry) to reduce our maintenance burden.

So next time when you come across an untested homemade util, before investing time to write tests for it, consider looking for a lib that does the same job and include it instead. Having to maintain less code is always preferable. :innocent:

Menardi commented on spec/unit/run.spec.js:
```diff
@@ -17,22 +17,217 @@
+
+            run.help();
+
+            expect(message.indexOf('Usage:') > -1);
```
Whoops, that's embarrassing!

Menardi commented on spec/unit/retry.spec.js:
```diff
@@ -0,0 +1,66 @@
+                .then(() => {
+                    done.fail('Unexpectedly resolved');
+                })
+                .fail(() => {
```
Ah I see! I specifically used Q rather than native Promises since I saw it being used everywhere else. I will use Promises from now on.

Menardi commented on spec/unit/retry.spec.js:
```diff
@@ -0,0 +1,66 @@
+
+            retry.retryPromise(attempts, promiseFn)
+                .then(() => {
+                    done.fail('Unexpectedly resolved');
```
I didn't realise Jasmine had a global `fail`, that is very convenient :+1:

Menardi commented:
Thanks for the detailed review, I really appreciate it. I would have looked into replacing `retryPromise` but I didn't want to be ripping things out in one of my first commits. I won't hesitate next time though :)

I have replied to most of your comments. On the topic of spying on `console.log`, I want to reiterate that Jasmine's behaviour is not what you might expect. I spent a while when originally writing the tests trying to figure out how to unspy, so that other tests would not be affected by the spy. But actually Jasmine unspies automatically after the test. I confirmed this by checking if `console.log` was spied in a following test, and it was not.

I am happy with your changes, and again very much appreciate your input. I will work on writing some more tests in another branch during the week.

By the way, I rebased the branch with master as there was a conflict in `package.json`.
